modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/dht/GridCachePartitionsStateValidationTest.java

```diff
@@ -323,4 +330,139 @@ public void waitUntilAllSingleMessagesAreSent() throws IgniteCheckedException {
             allSingleMessagesSent.get();
         }
     }
+
+    /**
+     * Three tests that partitions state validation works correctly and show partition size always (GG-24471):
+     *  1. Only partition counters are inconsistent.
+     *  2. Only partition size are inconsistent.
+     *  3. Partition counters and size are inconsistent.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testValidationIfPartitionCountersAreInconsistentWithSize() throws Exception {
+        testLogger = new ListeningTestLogger();
+
+        LogListener lsnr = LogListener.matches("Partitions update counters are inconsistent for ").build();
+
+        testLogger.registerListener(lsnr);
+
+        try {
+            IgniteEx ignite = startGrid(0);
+            startGrid(1);
+            startGrid(2);
+            ignite.cluster().active(true);
+            awaitPartitionMapExchange();
+
+            // Populate cache to increment update counters.
+            for (int i = 0; i < 1000; i++)
+                ignite.cache(CACHE_NAME).put(i, i);
+
+            // Modify update counter for some partition.
+            for (GridDhtLocalPartition partition : ignite.cachex(CACHE_NAME).context().topology().localPartitions()) {
+                partition.updateCounter(100500L);
+                break;
+            }
+
+            // Trigger exchange.
+            startGrid(3);
+
+            awaitPartitionMapExchange();
+
+            // Nothing should happen (just log error message) and we're still able to put data to corrupted cache.
+            ignite.cache(CACHE_NAME).put(0, 0);
+
+
+            assertTrue("Counters inconsistent message not found", lsnr.check());
+        }
+        finally {
+            stopAllGrids();
+        }
+    }
+
+    @Test
+    public void testValidationIfPartitionSizesAreInconsistentWithSize() throws Exception {
+
+        testLogger = new ListeningTestLogger();
+
+        LogListener lsnrSize = LogListener.matches("Partitions cache sizes are inconsistent").build();
+        testLogger.registerListener(lsnrSize);
+
+        try{
+            IgniteEx ignite = startGrid(0);
```

Comment:
In all three tests code inside try block looks very similar. I think it can be extracted to separate method.